### PR TITLE
Add scripts to perform split nulls

### PR DESCRIPTION
**Pull request by adrien-laposta**
This PR proposes to add three scripts to

- Compute the per-split noise power spectra
- Get the covariances between splits
- Compute residuals between cross-split spectra

I have also included a `use_per_split_beam` flag in the paramfile to add a per-split beam correction in `get_spectra_from_alms`.

### CODE REVIEW

File: project/data_analysis/python/get_spectra_from_alms.py [hunk 1]
```diff
@@ -130,6 +144,19 @@
...
+                beam_correction /= (bl_dict[sv1, ar1, s1] * bl_dict[sv2, ar2, s2])
+
+                if binned_mcm:
+                    _, beam_correction = pspy_utils.naive_binning(l, beam_correction,
+                                                                  binning_file, lmax)
+
+                for s in spin_pairs:
+                    if s == "spin2xspin2":
+                        mbb_inv[s] *= np.tile(beam_correction, 4)[:, None]
+                    mbb_inv[s] *= beam_correction[:, None]
```
Comment (xgarrido) [inline project/data_analysis/python/get_spectra_from_alms.py:158]: is there a `else` missing here ?

File: project/data_analysis/python/split_nulls/get_per_split_noise.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+        split_noise = {spec: split_ps_dict[i, i][spec] - split_noise_mean[spec] for spec in spectra}
+
+        split_noise_dict[sv1, ar1, sv2, ar2, i] = split_noise
+
+
+for (sv1, ar1, sv2, ar2, split), split_noise in split_noise_dict.items():
+    spec_name = f"Dl_{sv1}_{ar1}x{sv2}_{ar2}_{split}{split}_noise.dat"
+
+    so_spectra.write_ps(f"{split_noise_dir}/{spec_name}", lb, split_noise, type, spectra = spectra)
```
Comment (xgarrido) [inline project/data_analysis/python/split_nulls/get_per_split_noise.py:85]: pep8 recommands no space between `=` for function arguments

File: project/data_analysis/python/split_nulls/get_split_covariance_blocks.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+arrays = {sv: d[f"arrays_{sv}"] for sv in surveys}
+array_list = [f"{sv}_{ar}" for sv in surveys for ar in arrays[sv]]
+
+l_fg, fg_dict = best_fits.fg_dict_from_files(bestfit_dir + "/fg_{}x{}.dat", array_list, lmax, spectra)
+
+for sv in surveys:
+    arrays[sv] = d[f"arrays_{sv}"]
+    for ar in arrays[sv]:
+        l_beam, bl_dict[sv, ar] = pspy_utils.read_beam_file(d[f"beam_{sv}_{ar}"])
+        id_beam = np.where((l_beam >= 2) & (l_beam < lmax))
```
Comment (xgarrido) [inline project/data_analysis/python/split_nulls/get_split_covariance_blocks.py:53]: I'm not sure the `where` is needed here. I think only a slice selection is needed just like in the other files

File: project/data_analysis/python/split_nulls/get_split_nulls.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+for ar in arrays:
+
+    chi2_dict[ar] = {m: [] for m in modes}
+    pte_dict[ar] = {m: [] for m in modes}
+    ns = n_splits[ar]
+    splits_id = [str(i) for i in range(ns)]
+
+    cross_splits = list(combinations(splits_id, 2))
+    split_diff_list = list(combinations(cross_splits, 2))
+    print(split_diff_list)
```
Comment (xgarrido) [inline project/data_analysis/python/split_nulls/get_split_nulls.py:64]: I understand that for printing purpose you need to cast the `combinations` into `list` but for the rest of the code you do not need this cast. So I suggest `cross_splits` and `split_diff_list` remains `combinations` objects and only cast them for printing purpose. Not casting make the combinations faster.

**Review by xgarrido — APPROVED** [review]

**Comment by xzackli** [comment]
Excited to see this merge!

**Pull request merged**